src/Init/Data/BitVec/Lemmas.lean

```diff
@@ -1512,6 +1512,25 @@ theorem shiftRight_add {w : Nat} (x : BitVec w) (n m : Nat) :
   ext i
   simp [Nat.add_assoc n m i]
 
+@[simp]
+theorem shiftLeft_ushiftRight {x : BitVec w} {n : Nat} :
+  x >>> n <<< n = x &&& BitVec.allOnes w <<< n := by
+  induction n generalizing x
+  case zero =>
+    ext; simp
+  case succ n ih =>
+    rw [BitVec.shiftLeft_add, Nat.add_comm, BitVec.shiftRight_add, ih,
+       Nat.add_comm, BitVec.shiftLeft_add, BitVec.shiftLeft_and_distrib]
+    ext i
+    have hi₁ (_: 0 < i.val) : 1 + (i.val - 1) = i := by
+      rw [Nat.add_comm, Nat.sub_add_cancel]
+      omega
+    by_cases hw : w = 0
+    · simp [hw]
+    by_cases hi₂ : i.val = 0
+    · simp [hi₂]
+    · simp [Nat.lt_one_iff, hi₂, hi₁ (show 0 < i.val by omega)]
```

Comment:
AFAICT this is the only use of `hi₁` right? Given the fact that it and its premise are provable by `omega` I'd almost be certain that you can just put the `have` without the additional premise etc. in this specific case and prove it `by omega`, then rewrite with it in this `simp`.

Reply:
not sure I understood exactly what you mean, but without the premise (`have hi₁ : 1 + (i.val - 1) = i := by...`) `omega` is not sufficient (it is otherwise)

Reply:
Yes but if you prove the `have` in the context of the line that is marked here that hypothesis will be implicitly available so you should be able to directly show your fact as a `have` in this particular subgoal. Instead of proving the more general version above and then filling in the argument with another `by omega`